File: apps/main-service/src/routes/health.js
const express = require('express');
const axios = require('axios');
const { checkRedisHealth, cacheHelpers } = require('../config/redis');

const router = express.Router();

// 서비스 URL 설정
const CATALOG_SERVICE_URL = process.env.CATALOG_SERVICE_URL || 'http://catalog-service:8000';
const ORDER_SERVICE_URL = process.env.ORDER_SERVICE_URL || 'http://order-service:8080';

// 기본 헬스체크
router.get('/', async (req, res) => {
  const healthCheck = {
    service: 'Main Service',
    status: 'healthy',
    timestamp: new Date().toISOString(),
    uptime: process.uptime(),
    memory: process.memoryUsage(),
  };

  res.status(200).json(healthCheck);
});

// 상세 헬스체크 (모든 의존성 포함)
router.get('/detailed', async (req, res) => {
  const startTime = Date.now();
  
  try {
    // 각 서비스 헬스체크를 병렬로 실행
    const [catalogHealth, orderHealth, redisHealth] = await Promise.allSettled([
      checkServiceHealth(CATALOG_SERVICE_URL),
      checkServiceHealth(ORDER_SERVICE_URL),
      checkRedisHealth()
    ]);

    // Redis 통계 가져오기
    const redisStats = await cacheHelpers.getStats();
    const hitRate = redisStats ? 
      Math.round((redisStats.keyspace_hits / (redisStats.keyspace_hits + redisStats.keyspace_misses)) * 100) || 0 
      : 0;

    const healthData = {
      service: 'Main Service',
      status: 'healthy',
      timestamp: new Date().toISOString(),
      responseTime: Date.now() - startTime,
      services: [
        {
          name: 'Frontend',
          status: 'healthy', // Frontend는 정적 파일이므로 항상 healthy
          responseTime: 0
        },
        {
          name: 'Main Service',
          status: 'healthy',
          responseTime: Date.now() - startTime
        },
        {
          name: 'Catalog Service',
          status: catalogHealth.status === 'fulfilled' ? catalogHealth.value.status : 'unhealthy',
          responseTime: catalogHealth.status === 'fulfilled' ? catalogHealth.value.responseTime : 0
        },
        {
          name: 'Order Service',
          status: orderHealth.status === 'fulfilled' ? orderHealth.value.status : 'unhealthy',
          responseTime: orderHealth.status === 'fulfilled' ? orderHealth.value.responseTime : 0
        }
      ],
      database: {
        status: 'connected', // 실제로는 각 서비스에서 확인
        connections: 12 // 목업 데이터
      },
      cache: {
        status: redisHealth.status,
        hitRate: hitRate,
        connections: redisStats ? redisStats.connected_clients : 0,
        memory: redisStats ? Math.round(redisStats.used_memory / 1024 / 1024) : 0 // MB
      },
      storage: {
        status: 'available', // S3는 별도 헬스체크 필요
        usage: 34 // 목업 데이터
      }
    };

    res.status(200).json(healthData);
  } catch (error) {
    console.error('Health check error:', error);
    res.status(503).json({
      service: 'Main Service',
      status: 'unhealthy',
      timestamp: new Date().toISOString(),
      error: error.message
    });
  }
});

// 개별 서비스 헬스체크 함수
async function checkServiceHealth(serviceUrl) {
  const startTime = Date.now();
  
  try {
    const response = await axios.get(`${serviceUrl}/health`, {
      timeout: 5000,
      headers: {
        'User-Agent': 'Main-Service-HealthCheck/1.0'
      }
    });
    
    return {
      status: response.status === 200 ? 'healthy' : 'unhealthy',
      responseTime: Date.now() - startTime
    };
  } catch (error) {
    console.error(`Health check failed for ${serviceUrl}:`, error.message);
    return {
      status: 'unhealthy',
      responseTime: Date.now() - startTime,
      error: error.message
    };
  }
}

// Kubernetes liveness probe
router.get('/live', (req, res) => {
  res.status(200).send('OK');
});

// Kubernetes readiness probe
router.get('/ready', async (req, res) => {
  try {
    // Redis 연결 확인
    const redisHealth = await checkRedisHealth();
    
    if (redisHealth.status === 'connected') {
      res.status(200).send('Ready');
    } else {
      res.status(503).send('Not Ready - Redis unavailable');
    }
  } catch (error) {
    res.status(503).send('Not Ready - Health check failed');
  }
});

// 메트릭 엔드포인트 (Prometheus 형식)
router.get('/metrics', async (req, res) => {
  try {
    const redisStats = await cacheHelpers.getStats();
    const memoryUsage = process.memoryUsage();
    
    const metrics = `
# HELP main_service_uptime_seconds Total uptime of the service in seconds
# TYPE main_service_uptime_seconds counter
main_service_uptime_seconds ${process.uptime()}

# HELP main_service_memory_usage_bytes Memory usage in bytes
# TYPE main_service_memory_usage_bytes gauge
main_service_memory_usage_bytes{type="rss"} ${memoryUsage.rss}
main_service_memory_usage_bytes{type="heapTotal"} ${memoryUsage.heapTotal}
main_service_memory_usage_bytes{type="heapUsed"} ${memoryUsage.heapUsed}

# HELP redis_keyspace_hits_total Total number of successful lookups of keys
# TYPE redis_keyspace_hits_total counter
redis_keyspace_hits_total ${redisStats ? redisStats.keyspace_hits : 0}

# HELP redis_keyspace_misses_total Total number of failed lookups of keys
# TYPE redis_keyspace_misses_total counter
redis_keyspace_misses_total ${redisStats ? redisStats.keyspace_misses : 0}

# HELP redis_connected_clients Number of client connections
# TYPE redis_connected_clients gauge
redis_connected_clients ${redisStats ? redisStats.connected_clients : 0}
    `.trim();

    res.set('Content-Type', 'text/plain');
    res.send(metrics);
  } catch (error) {
    console.error('Metrics error:', error);
    res.status(500).send('Error generating metrics');
  }
});

module.exports = router;

File: apps/catalog-service/src/routers/health.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from datetime import datetime
import logging

from ..database import get_db, test_database_connection
from ..cache import cache_manager
from ..schemas import HealthResponse, DetailedHealthResponse
from ..config import settings

router = APIRouter()
logger = logging.getLogger(__name__)

@router.get("/", response_model=HealthResponse)
async def health_check():
    """기본 헬스체크"""
    return HealthResponse(
        service="Catalog Service",
        status="healthy",
        timestamp=datetime.utcnow(),
        version="1.0.0",
        environment=settings.ENVIRONMENT
    )

@router.get("/detailed", response_model=DetailedHealthResponse)
async def detailed_health_check(db: AsyncSession = Depends(get_db)):
    """상세 헬스체크"""
    
    # 데이터베이스 연결 확인
    db_healthy = await test_database_connection()
    db_status = {
        "status": "connected" if db_healthy else "disconnected",
        "url": settings.DATABASE_URL.split("@")[-1] if "@" in settings.DATABASE_URL else "unknown"
    }
    
    # Redis 연결 확인
    cache_healthy = await cache_manager.ping()
    cache_stats = await cache_manager.get_stats() if cache_healthy else {}
    cache_status = {
        "status": "connected" if cache_healthy else "disconnected",
        "stats": cache_stats
    }
    
    # 의존성 서비스 상태 (실제로는 다른 서비스들과의 연결 확인)
    dependencies = {
        "main_service": {
            "status": "unknown",
            "description": "Main API Gateway Service"
        }
    }
    
    return DetailedHealthResponse(
        service="Catalog Service",
        status="healthy" if db_healthy and cache_healthy else "degraded",
        timestamp=datetime.utcnow(),
        version="1.0.0",
        environment=settings.ENVIRONMENT,
        database=db_status,
        cache=cache_status,
        dependencies=dependencies
    )

@router.get("/live")
async def liveness_probe():
    """Kubernetes Liveness Probe"""
    return {"status": "alive"}

@router.get("/ready")
async def readiness_probe(db: AsyncSession = Depends(get_db)):
    """Kubernetes Readiness Probe"""
    
    # 데이터베이스 연결 확인
    db_healthy = await test_database_connection()
    
    # Redis 연결 확인
    cache_healthy = await cache_manager.ping()
    
    if db_healthy and cache_healthy:
        return {"status": "ready"}
    else:
        from fastapi import HTTPException
        raise HTTPException(status_code=503, detail="Service not ready")

@router.get("/metrics")
async def metrics():
    """Prometheus 메트릭"""
    
    # 캐시 통계
    cache_stats = await cache_manager.get_stats()
    
    # 기본 메트릭들
    metrics_data = f"""
# HELP catalog_service_uptime_seconds Total uptime of the service in seconds
# TYPE catalog_service_uptime_seconds counter
catalog_service_uptime_seconds {cache_stats.get('uptime_in_seconds', 0)}

# HELP catalog_service_cache_hits_total Total number of cache hits
# TYPE catalog_service_cache_hits_total counter
catalog_service_cache_hits_total {cache_stats.get('keyspace_hits', 0)}

# HELP catalog_service_cache_misses_total Total number of cache misses
# TYPE catalog_service_cache_misses_total counter
catalog_service_cache_misses_total {cache_stats.get('keyspace_misses', 0)}

# HELP catalog_service_cache_memory_usage_bytes Cache memory usage in bytes
# TYPE catalog_service_cache_memory_usage_bytes gauge
catalog_service_cache_memory_usage_bytes {cache_stats.get('used_memory', 0)}

# HELP catalog_service_cache_connected_clients Number of connected cache clients
# TYPE catalog_service_cache_connected_clients gauge
catalog_service_cache_connected_clients {cache_stats.get('connected_clients', 0)}

# HELP catalog_service_commands_processed_total Total number of commands processed
# TYPE catalog_service_commands_processed_total counter
catalog_service_commands_processed_total {cache_stats.get('total_commands_processed', 0)}
    """.strip()
    
    from fastapi import Response
    return Response(content=metrics_data, media_type="text/plain")